libraries/AP_ExternalAHRS/AP_ExternalAHRS_GSOF.cpp

```diff
@@ -0,0 +1,362 @@
+                        gps_data.fix_type = AP_GPS_FixType::NONE;
+                        break;
+                }
+                if (ins_full_nav.gnss_status != AP_GSOF::GnssStatus::FIX_NOT_AVAILABLE) {
+                    // If fix is unavailble, the lat/lon may be zero, so don't post the data.
+                    // To reproduce this condition, disconnect the GPS antenna and reboot the receiver.
+                    post_filter();
+                }
+            }
+            if (parsed.get(AP_GSOF::INS_RMS)) {
+                last_ins_rms_ms = now;
+
+                gps_data.horizontal_pos_accuracy = Vector2d(ins_rms.pos_rms_n, ins_rms.pos_rms_e).length();
+                gps_data.vertical_pos_accuracy = ins_rms.pos_rms_d;
+                gps_data.horizontal_vel_accuracy = Vector2d(ins_rms.vel_rms_n, ins_rms.vel_rms_e).length();
+            }
+            if (parsed.get(AP_GSOF::LLH_MSL)) {
+                last_llh_msl_ms = now;
+
+                gps_data.longitude = static_cast<int32_t>(llh_msl.longitude * 1E7);
+                gps_data.latitude = static_cast<int32_t>(llh_msl.latitude * 1E7);
+                gps_data.msl_altitude = static_cast<int32_t>(llh_msl.altitude_msl * 1E2);
+            }
+
+            if (parsed.get(AP_GSOF::LLH_MSL) && parsed.get(AP_GSOF::INS_FULL_NAV)) {
+                undulation = ins_full_nav.altitude - llh_msl.altitude_msl;
+            }
+
+            uint8_t instance;
+            AP_GSOF::MsgTypes expected_gps;
+            expected_gps.set(AP_GSOF::POS_TIME);
+            expected_gps.set(AP_GSOF::INS_FULL_NAV);
+            expected_gps.set(AP_GSOF::INS_RMS);
+            expected_gps.set(AP_GSOF::LLH_MSL);
+            if (AP::gps().get_first_external_instance(instance) && parsed == expected_gps) {
+                AP::gps().handle_external(gps_data, instance);
+            }
+        }
+        hal.scheduler->delay_microseconds(100);
+        check_initialise_state();
+    }
+}
+
+void AP_ExternalAHRS_GSOF::check_initialise_state(void)
+{
+    const auto new_init_state = initialised();
+    // Only send the message after fully booted up, otherwise it gets dropped.
+    if (!last_init_state && new_init_state && AP_HAL::millis() > 5000) {
+        GCS_SEND_TEXT(MAV_SEVERITY_INFO, LOG_FMT, get_name(), "initialised.");
+        last_init_state = new_init_state;
+    }
+}
+
+void AP_ExternalAHRS_GSOF::post_filter() const
+{
+    WITH_SEMAPHORE(state.sem);
+    state.velocity = Vector3f{ins_full_nav.vel_n, ins_full_nav.vel_e, ins_full_nav.vel_d};
+    state.have_velocity = true;
+
+    state.location = Location(
+        ins_full_nav.latitude * 1E7,
+        ins_full_nav.longitude * 1E7,
+        (ins_full_nav.altitude - undulation) * 1E2,
+        Location::AltFrame::ABSOLUTE);
+    state.have_location = true;
+
+    state.quat.from_euler(
+        radians(ins_full_nav.roll_deg), 
+        radians(ins_full_nav.pitch_deg),
+        radians(wrap_360(ins_full_nav.heading_deg)));
```

Comment:
Well this is annoying, even with wrap_360, you still get this result:
![image](https://github.com/user-attachments/assets/e5bfb804-f2bd-47e3-9ce9-72a5f07ba828)